thorlcr/activities/nsplitter/thnsplitterslave.cpp

```diff
@@ -194,55 +196,63 @@ class NSplitterSlaveActivity : public CSlaveActivity, implements ISharedSmartBuf
-                    smartBuf.setown(createSharedSmartDiskBuffer(this, tempname.str(), numOutputs, queryRowInterfaces(input), &container.queryJob().queryIDiskUsage()));
-                    ActPrintLog("Using temp spill file: %s", tempname.str());
+                    CSplitterOutput *output = (CSplitterOutput *)outputs.item(o);
+                    if (output->isStopped())
+                        --activeOutputCount;
                 }
+                assertex(activeOutputCount); // must be >=1, as this output (outIdx) has invoked prepareInput
+                if (1 == activeOutputCount)
+                    return false;
+                if (smartBuf)
+                    smartBuf->reset();
                 else
                 {
-                    ActPrintLog("Spill is 'balanced'");
-                    smartBuf.setown(createSharedSmartMemBuffer(this, numOutputs, queryRowInterfaces(input), NSPLITTER_SPILL_BUFFER_SIZE));
-                }
-                // mark any outputs already stopped
-                ForEachItemIn(o, outputs)
-                {
-                    CSplitterOutput *output = (CSplitterOutput *)outputs.item(o);
-                    if (output->isStopped() || !connectedOutputSet->test(o))
-                        smartBuf->queryOutput(o)->stop();
+                    if (spill)
+                    {
+                        StringBuffer tempname;
+                        GetTempName(tempname, "nsplit", true); // use alt temp dir
+                        smartBuf.setown(createSharedSmartDiskBuffer(this, tempname.str(), numOutputs, queryRowInterfaces(input), &container.queryJob().queryIDiskUsage()));
+                        ActPrintLog("Using temp spill file: %s", tempname.str());
+                    }
+                    else
+                    {
+                        ActPrintLog("Spill is 'balanced'");
+                        smartBuf.setown(createSharedSmartMemBuffer(this, numOutputs, queryRowInterfaces(input), NSPLITTER_SPILL_BUFFER_SIZE));
+                    }
+                    // mark any outputs already stopped
+                    ForEachItemIn(o, outputs)
+                    {
+                        CSplitterOutput *output = (CSplitterOutput *)outputs.item(o);
+                        if (output->isStopped() || !connectedOutputSet->test(o))
+                            smartBuf->queryOutput(o)->stop();
+                    }
                 }
+                if (!spill)
+                    writer.start(); // writer keeps writing ahead as much as possible, the readahead impl. will block when has too much
+            }
+            catch (IException *e)
+            {
+                eofHit = true;
+                writeAheadException.setown(e);
+                return false;
             }
-            if (!spill)
-                writer.start(); // writer keeps writing ahead as much as possible, the readahead impl. will block when has too much
         }
         return true;
```

Comment:
@jakesmith  Should this be returning false if the first instance failed to start?  Should it be using a different exception member to distinguish between a problem starting and a problem when reading rows.

Reply:
>Should this be returning false if the first instance failed to start?

The failure in parepareInput triggers eofHit and any subsequent caller to writeahead will exit at start.

> Should it be using a different exception member to distinguish between a problem starting and a problem when reading rows.

I considered it, but a failure on start is a failure at 0 == recsReady, i.e. the condition met by the common test in NSplitterSlaveActivity::nextRow, so think it's cleaner like it is.